File: C/test_roman_numerals.c
/*
 * Test Suite for Roman Numeral Converter
 * Copyright © MMXVI-MMXVII, Chris Warrick.
 * All rights reserved.
 * License: 3-clause BSD, see main file for details.
 */

#include "roman_numerals.c"
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

int test_tr(int integer, char* roman) {
    char *output = (char*)malloc(16);
    to_roman(integer, output);
    int ret;
    if (strcmp(output, roman) == 0) {
        putchar('.');
        ret = 0;
    } else {
        putchar('F');
        ret = 1;
    }
    free(output);
    return ret;
}

int test_tr_error(int integer) {
    char *output = (char*)malloc(16);
    int val = to_roman(integer, output);
    int ret;
    if (val == -1) {
        putchar('.');
        ret = 0;
    } else {
        putchar('F');
        ret = 1;
    }
    free(output);
    return ret;

}


int test_fr(int integer, char* roman) {
    int output = from_roman(roman);
    if (output == integer) {
        putchar('.');
        return 0;
    } else {
        putchar('F');
        return 1;
    }
}

int main() {
    FILE* fh = fopen("../test_data.txt", "r");
    char* line = (char*)malloc(32);
    char *output_to = (char*)malloc(16);
    int output_from;
    int integer;
    char* roman;
    int errors = 0;
    char* token;

    // edge cases test
    printf("to_roman");
    errors += test_tr(1, "I");
    errors += test_tr(2, "II");
    errors += test_tr(3, "III");
    errors += test_tr(4, "IV");
    errors += test_tr(5, "V");
    errors += test_tr(6, "VI");
    errors += test_tr(1234, "MCCXXXIV");
    errors += test_tr(1958, "MCMLVIII");
    errors += test_tr(2222, "MMCCXXII");
    errors += test_tr(3999, "MMMCMXCIX");

    printf("\nto_roman errors");
    test_tr_error(0);
    test_tr_error(-1);
    test_tr_error(4000);

    printf("\nfrom_roman");
    test_fr(1, "I");
    test_fr(2, "Ii");
    test_fr(3, "iii");
    test_fr(4, "  iv  ");
    test_fr(5, "V");
    test_fr(8, "ViIi");

    // file tests
    printf("\nall inputs...");
    while (fgets(line, 32, fh) != NULL) {
        token = strtok(line, " ");
        integer = atoi(token);
        roman = strtok(NULL, " ");
        // remove \n
        roman = strtok(roman, "\n");

        // make buffer and test
        to_roman(integer, output_to);
        if (strcmp(roman, output_to) != 0) {
            printf("ERROR: to_roman, expected %s, received %s\n", roman, output_to);
            errors++;
        }

        output_from = from_roman(roman);
        if (output_from != integer) {
            printf("ERROR: from_roman, expected %d, received %d\n", integer, output_from);
            errors++;
        }
    }
    printf("\n");
    if (errors == 0) puts("PASS");
    else printf("%d ERRORS\n", errors);
    free(line);
    free(output_to);
    return errors != 0;
}

File: Lua/test_roman_numerals.lua
-- Test Suite for Roman Numeral Converter
-- Copyright © MMXVI-MMXVIII, Chris Warrick.
-- All rights reserved.
-- License: 3-clause BSD, see main file for details.

require('roman_numerals')
local lu = require('luaunit')

TestRomanNumerals = {} -- class
    function test_to_roman()
        lu.assertEquals(to_roman(1), "I")
        lu.assertEquals(to_roman(2), "II")
        lu.assertEquals(to_roman(3), "III")
        lu.assertEquals(to_roman(4), "IV")
        lu.assertEquals(to_roman(5), "V")
        lu.assertEquals(to_roman(6), "VI")
        lu.assertEquals(to_roman(1234), "MCCXXXIV")
        lu.assertEquals(to_roman(1958), "MCMLVIII")
        lu.assertEquals(to_roman(2222), "MMCCXXII")
        lu.assertEquals(to_roman(3999), "MMMCMXCIX")
    end

    function test_to_roman_invalid()
        lu.assertErrorMsgContains("Integer must be between 1 and 3999", to_roman, 0)
        lu.assertErrorMsgContains("Integer must be between 1 and 3999", to_roman, -1)
        lu.assertErrorMsgContains("Integer must be between 1 and 3999", to_roman, 4000)
    end

    function test_from_roman()
        lu.assertEquals(from_roman("I"), 1)
        lu.assertEquals(from_roman("Ii"), 2)
        lu.assertEquals(from_roman("iii"), 3)
        lu.assertEquals(from_roman("  iv  "), 4)
        lu.assertEquals(from_roman("V"), 5)
        lu.assertEquals(from_roman("ViIi"), 8)
    end

    function test_to_roman_file()
        local fh = assert(io.open('../test_data.txt'))

        while true do
            local line = fh:read("*line")
            if not line then break end
            local space = line:find(" ")
            integer = tonumber(line:sub(1, space))
            integer = tonumber(integer)
            roman = line:sub(space, -1)
            lu.assertEquals(to_roman(integer), roman)
        end
        fh:close()
    end

    function test_from_roman_file()
        local fh = assert(io.open('../test_data.txt'))

        while true do
            local line = fh:read("*line")
            if not line then break end
            local space = line:find(" ")
            integer = tonumber(line:sub(1, space))
            integer = tonumber(integer)
            roman = line:sub(space, -1)
            lu.assertEquals(from_roman(roman), integer)
        end
        fh:close()
    end
-- class TestRomanNumerals

local runner = lu.LuaUnit.new()
runner:setOutputType("tap")
os.exit( runner:runSuite() )
